contrib/golang/filters/http/source/golang_filter.cc

```diff
@@ -1242,29 +1244,114 @@ FilterConfig::FilterConfig(
+  config_->plugin_name_ptr = name_ptr;
+  config_->plugin_name_len = plugin_name_.length();
+  config_->config_ptr = buf_ptr;
+  config_->config_len = buf.length();
+  config_->is_route_config = 0;
+
+  config_id_ = dso_lib_->envoyGoFilterNewHttpPluginConfig(config_);
+
   if (config_id_ == 0) {
-    throw EnvoyException(fmt::format("golang filter failed to parse plugin config: {} {}",
-                                     proto_config.library_id(), proto_config.library_path()));
+    throw EnvoyException(
+        fmt::format("golang filter failed to parse plugin config: {} {}", so_id_, so_path_));
   }
+
   ENVOY_LOG(debug, "golang filter new plugin config, id: {}", config_id_);
-};
+}
 
 FilterConfig::~FilterConfig() {
   if (config_id_ > 0) {
     dso_lib_->envoyGoFilterDestroyHttpPluginConfig(config_id_);
   }
 }
 
+CAPIStatus FilterConfig::defineMetric(uint32_t metric_type, absl::string_view name,
+                                      uint32_t* metric_id) {
+  Thread::LockGuard lock(mutex_);
+  if (metric_type > static_cast<uint32_t>(MetricType::Max)) {
+    return CAPIStatus::CAPIValueNotFound;
+  }
+
+  auto type = static_cast<MetricType>(metric_type);
+
+  Stats::StatNameManagedStorage storage(name, metric_store_->scope_->symbolTable());
+  Stats::StatName stat_name = storage.statName();
+  if (type == MetricType::Counter) {
+    auto id = metric_store_->nextCounterMetricId();
+    auto c = &metric_store_->scope_->counterFromStatName(stat_name);
+    metric_store_->counters_.emplace(id, c);
+    *metric_id = id;
+  } else if (type == MetricType::Gauge) {
+    auto id = metric_store_->nextGaugeMetricId();
+    auto g =
+        &metric_store_->scope_->gaugeFromStatName(stat_name, Stats::Gauge::ImportMode::Accumulate);
+    metric_store_->gauges_.emplace(id, g);
+    *metric_id = id;
+  } else { // (type == MetricType::Histogram)
+    ASSERT(type == MetricType::Histogram);
+    auto id = metric_store_->nextHistogramMetricId();
+    auto h = &metric_store_->scope_->histogramFromStatName(stat_name,
+                                                           Stats::Histogram::Unit::Unspecified);
+    metric_store_->histograms_.emplace(id, h);
+    *metric_id = id;
+  }
+
+  return CAPIStatus::CAPIOK;
+}
+
+CAPIStatus FilterConfig::incrementMetric(uint32_t metric_id, int64_t offset) {
+  Thread::LockGuard lock(mutex_);
+  auto type = static_cast<MetricType>(metric_id & MetricStore::kMetricTypeMask);
+  if (type == MetricType::Counter) {
+    auto it = metric_store_->counters_.find(metric_id);
+    if (it != metric_store_->counters_.end()) {
+      if (offset > 0) {
+        it->second->add(offset);
+      }
+    }
+  } else if (type == MetricType::Gauge) {
+    auto it = metric_store_->gauges_.find(metric_id);
+    if (it != metric_store_->gauges_.end()) {
+      if (offset > 0) {
+        it->second->add(offset);
+      } else {
+        it->second->sub(-offset);
+      }
+    }
+  }
+  return CAPIStatus::CAPIOK;
```

Comment:
we just ignore Histogram?

Reply:
maybe Histogram has no add interface? it can be update with record interface in `recordMetric` to be implemented.